substrate/frame/distribution/src/functions.rs

```diff
@@ -0,0 +1,121 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helper functions for Distribution pallet.
+
+pub use super::*;
+impl<T: Config> Pallet<T> {
+	pub fn pot_account() -> AccountIdOf<T> {
+		// Get Pot account
+		let pot_id = T::PotId::get();
+		pot_id.into_account_truncating()
+	}
+
+	/// Funds transfer from the Pot to a project account
+	pub fn spend(amount: BalanceOf<T>, beneficiary: AccountIdOf<T>) -> DispatchResult {
+		// Get Pot account
+		let pot_account: AccountIdOf<T> = Self::pot_account();
+
+		//Operate the transfer
+		T::NativeBalance::transfer(&pot_account, &beneficiary, amount, Preservation::Preserve)?;
+
+		Ok(())
+	}
+
+	/// Series of checks on the Pot, to ensure that we have enough funds
+	/// before executing a Spend --> used in tests.
+	pub fn pot_check(spend: BalanceOf<T>) -> DispatchResult {
+		// Get Pot account
+		let pot_account = Self::pot_account();
+
+		// Check that the Pot as enough funds for the transfer
+		let balance = T::NativeBalance::balance(&pot_account);
+		let minimum_balance = T::NativeBalance::minimum_balance();
+		let remaining_balance = balance.saturating_sub(spend);
+
+		ensure!(remaining_balance > minimum_balance, Error::<T>::InsufficientPotReserves);
+		ensure!(balance > spend, Error::<T>::InsufficientPotReserves);
+		Ok(())
+	}
+
+	// Done in begin_block
+	// At the beginning of every Epoch, populate the `Spends` storage from the `Projects` storage
+	// (populated by an external process/pallet) make sure that there is enough funds before
+	// creating a new `SpendInfo`, and `ProjectInfo` corresponding to a created `SpendInfo`
+	// should be removed from the `Projects` storage. This is also a good place to Reserve the
+	// funds for created `SpendInfos`. the function will be use in a hook.
+
+	pub fn begin_block(now: BlockNumberFor<T>) -> Weight {
+		let max_block_weight = T::BlockWeights::get().max_block / 4;
```

Comment:
This cannot really work once the govenrance is on AH, as the we cannot afford to use too much of the block space anymore.. 

In general, in light of moving the governance to AH, it is best to design this pallet as lazy as possible, ideally without using on_initialize: All operations should be triggered via transactions.

Reply:
@kianenigma , is there the same concern when using `on_idle` instead of `on_initialize` ?